### PR TITLE
Scan only PR commits for Gitleaks instead of whole codebase

### LINKED ISSUE #2487: Scan only PR commits for Gitleaks instead of whole codebase
### Is your feature request related to a problem? Please describe.

Scan only PR commits for Gitleaks instead of whole codebase, when `VALIDATE_ALL_CODEBASE` is set to `false`

### Describe the solution you'd like

When `VALIDATE_ALL_CODEBASE` is set to `false`, and megalinter runs on PR then Gitleaks scans only PR commits.

### Describe alternatives you've considered

Described scenarios is achievable by passing `REPOSITORY_GITLEAKS_ARGUMENTS` env contains gitleaks flag `--log-opts`, but nice to have out-of-the-box support (using `VALIDATE_ALL_CODEBASE`) without extra configuration.

```yaml
- name: MegaLinter
  if: ${{ github.event_name == 'pull_request' }}
  id: megalinter
  uses: oxsecurity/megalinter/flavors/documentation@v6
  env:
    VALIDATE_ALL_CODEBASE: false
    REPOSITORY_GITLEAKS_ARGUMENTS: --log-opts '--no-merges --first-parent ${{ github.event.pull_request.base.sha }}^..${{ github.event.pull_request.head.sha }}'
```

Downside - to scan only PR commits with proposed gitleaks config, git checkout has to be set with `fetch-depth: 0`

```yaml
- name: Checkout
  uses: actions/checkout@v3
  with:
    fetch-depth: 0
```

### Additional context

N/A

### PR DESCRIPTION
**Pull request by DariuszPorowski**
Fixes #2487 

## Proposed Changes

1. `--redact` flag for gitleaks should be default to prevent exposing detected secrets to logs
2. common functions consolidated into utils, based on review comment: https://github.com/oxsecurity/megalinter/pull/2504#discussion_r1155263452
3. `CI` env use case, based on discussion: https://github.com/oxsecurity/megalinter/pull/2510#discussion_r1155338496
4. Missed doc fixes in https://github.com/oxsecurity/megalinter/pull/2449 (fixes https://github.com/oxsecurity/megalinter/issues/2448)
5. updated url to new gitleaks home: `https://github.com/zricethezav/gitleaks` -> `https://github.com/gitleaks/gitleaks`

github

ado

gitlab

### Author/Contributor
- [x] Add entry to the [CHANGELOG](https://github.com/oxsecurity/megalinter/blob/main/CHANGELOG.md) listing the change and linking to the corresponding issue (if appropriate)
- [x] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

**Comment by echoix** [comment]
For the `pyproject.toml`, you can keep it for another PR, it's already in the plans to switch to it, but neither of us had time to learn about it yet! If you are used to it, we'll gladly accept a first implementation  😄    
When building, there is a warning saying pip version 23 will enforce it, ie something like 

> DEPRECATION: xxxxxxx is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at pypa/pip#8559


So we'll need to go to it a way or another.

**Comment by DariuszPorowski** [comment]
@echoix that's the reason I put the note in the 1st line `NOTE: pyproject.toml and poetry is just for my dev env, will remove on ready PR` - so do not worry about these files for now :)
I can prapare PR with pyproject.toml and poetry in the future, poetry with venv changed my life ;)

**Comment by Kurt-von-Laven** [comment]
Yeah, Poetry is great!

File: megalinter/linters/GitleaksLinter.py [hunk 1]
```diff
@@ -13,48 +13,47 @@
...
 class GitleaksLinter(Linter):
     def __init__(self, params=None, linter_config=None):
-        self.is_pr, self.pr_source_sha, self.pr_target_sha = self.get_pr_data()
         super().__init__(params, linter_config)
+        self.only_pr_commits = config.get("REPOSITORY_GITLEAKS_ONLY_PR_COMMITS", "false")
+        if self.only_pr_commits == "true":
+            self.is_pr, self.pr_source_sha, self.pr_target_sha = self.get_pr_data()
+
 
     def get_pr_data(self):
```
Comment (nvuillam) [inline megalinter/linters/GitleaksLinter.py:22]: I'm pretty sure that such code could be reusable elsewhere by other linters, reporters or even core stuff, maybe you could put it in a new utils class ?

Comment (DariuszPorowski) [reply megalinter/linters/GitleaksLinter.py:22]: @nvuillam That was my initial thought and looked for utils class :) and realized it does not exist, and only utils for reporters is. Sure, will move the code.

**Comment by DariuszPorowski** [comment]
@nvuillam +/- this PR is ready, minor cleanup, and changes left, but have Q around `REPOSITORY_GITLEAKS_PR_COMMITS_SCAN`. Currently, it is false but happy to put true as default because then validate all code base = false will does the job without any extra envs. But this is a kind of braking change and may impact current users' setup.
So, I'm happy to hear feedback and design decisions.

**Comment by nvuillam** [comment]
@DariuszPorowski it looks great, i'll check that tomorrow, thanks for the PR:)

**Comment by nvuillam** [comment]
Cc @Kurt-von-Laven @bdovaz @echoix

**Comment by DariuszPorowski** [comment]
@nvuillam just following up :)

**Comment by nvuillam** [comment]
@DariuszPorowski you have many failing test cases ^^

**Comment by nvuillam** [comment]
@DariuszPorowski if it breaks existing config of ML using gitleaks and VALIDATE_ALL_CODEBASE=false , i'd prefer to stay deactivated by default, to avoid a breaking change like u said ^^

**Comment by DariuszPorowski** [comment]
@nvuillam I'm sorry it took so long but the last month has been hectic at work. PR ready for review :)

**Comment by nvuillam** [comment]
@DariuszPorowski same here, I understand 🤣 

Your PR looks great but I have some doubts about the documentation... it seems you directly updated generated markdown file https://github.com/oxsecurity/megalinter/blob/cff15a01ef2d29d904b97b268df8b54225429b4a/docs/descriptors/repository_gitleaks.md , so the next time the doc will be rebuilt, it will be overwritten

If you want "free text" doc to be added, you need to update linter_description property in the YML descriptor file :)

**Comment by nvuillam** [comment]
@DariuszPorowski  Same remark about custom variables: if you add them in the descriptor, they will be automatically added to the doc & json schema ^^

image

**Comment by DariuszPorowski** [comment]
@nvuillam Oh gosh, the last changes to te main were painful to integrate with this PR :D hope now is good, docs generated automatically

**Comment by nvuillam** [comment]
@DariuszPorowski i can assure you they were even harder to implement 😅

But it's for the better good: with next version you'll just have to trust MegaLinter core code with your env variables (including secrets defined in CI/CD), no need anymore to trust the dozens of embedded linters 😎

**Comment by nvuillam** [comment]
@DariuszPorowski plz can u merge main in your branch ? :)

**Comment by DariuszPorowski** [comment]
Sure, will do over the weekend

**Comment by nvuillam** [comment]
I did it :)

**Review by nvuillam — COMMENTED** [review]
@DariuszPorowski thanks for this great PR :)

**Pull request merged**